src/libawkward/forth/ForthMachine.cpp

```diff
@@ -1256,9 +1270,13 @@ namespace awkward {
             "call 'begin' to 'step' again (note: check 'is_done')");
         }
         case util::ForthError::user_halt: {
-          throw std::invalid_argument(
-            "'user halt' in AwkwardForth runtime: user-defined error or stopping "
-            "condition");
+          std::string error_message("'user halt' in AwkwardForth runtime: user-defined error or stopping "
+          "condition");
+          if (stack_can_pop()) {
+            error_message = string_at(stack().back()).empty() ? error_message :
+              string_at(stack().back());
+          }
+          throw std::invalid_argument(error_message);
```

Comment:
Let's not change `maybe_throw`. I think it would be better if the `util::ForthError::user_halt` exception does not require Forth programs to be written in such a way that they always set a string immediately before the `halt` keyword.

The above code assumes that `halt` is only ever used this way:

```forth
s" Error message" halt
```

and without the `s" Error message"`, it would be a C++ segfault (string not found in `strings_`) or a confusing behavior: returning a random string as an error message, if that string is defined anywhere else in the Forth code and there happened to be the right integer on the top of the stack.

Let's have the assumption that `s" Error message"` always precedes `halt` be an assumption that only TypedArrayBuilder makes and guarantees.

Reply:
done